_published/0012-authentication.md

```diff
@@ -0,0 +1,167 @@
+with PR number/ID and feature name. Use leading zeros rfc_author_username:
+fredx30 rfc_author_name: Fredrik Dyrvold # Or same as username, if you wish
+
+# Leave these. Collaborator changes this before merging
+
+impl_issue_id: 0 impl_issue_repo: iver-wharf/wharf-api last_modified_date:
+2021-06-16
+---
+
+# {{page.title}}
+
+- RFC
+  PR: [iver-wharf/rfcs#{{page.pr_id}}](https://github.com/iver-wharf/rfcs/pulls/{{page.pr_id}})
+- Feature name: `{{page.rfc_feature_name}}`
+- Author:
+  {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation
+  issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation
+  status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Wharf should have authentication. Wharf currently has no authentication. This
+will outline a plan to implement [OpenID Connect (OICD)](https://openid.net/connect/)
+in wharf. This RFC will only outline authentication with OICD and not
+authorization and user/group
+management. A further plan to implement user groups and authorization is
+expected to be drawn up in a later document.
+
+## Motivation
+
+Some claim a good feature of secure infrastructure is access control. Currently,
+our only solution is network isolation. Being able to restrict access to varying
+degrees is a product goal to satisfy business needs.
+
+## Explanation
+
+Pre implementation stages
+
+1. OICD in any form demands full SSL security in all connections. In order to
+   avoid having to fill the development environment with --force-unsecure
+   bindings we should consider implementing certificate generation and importing
+   for dev environments.
+
+Implementations stages
+
+1. Docker compose setup for Hydra Proxy setup
+   - This involves integration with the currently existing DB.
+2. Hydra (Authorisation server) - Connect to IDP (Config for Azure AD iver.se)
+   - Config instructions documented [here](https://www.ory.sh/hydra/docs/reference/configuration)
+   - [OpenID-Configuration](https://login.microsoftonline.com/common/.well-known/openid-configuration)
+   -
+3. Wharf-web - configure and setup [angular-auth-oicd-client](https://github.com/damienbod/angular-auth-oidc-client)
+   - Implement login endpoint
+   - Implement consent endpoint
+   - Implement logout endpoint
+
+4. Wharf-api - Implement and test
+   - [Go SDK for ory Hydra](https://www.ory.sh/hydra/docs/sdk/go)
+   - Inspiration for [gin middleware here](https://github.com/janekolszak/gin-hydra/blob/master/ginhydra.go)
+     note this is outdated.
+
+7. Edit wharf-web services to use bearer tokens through angular middleware
+   configured through `agular-auth-oicd-client`
+
+8. Wharf-provider-* - Implement and test
+   - Likely identical to procedure for wharf-api.
```

Comment:
The providers today only forwards the "Authorization" header. If it receives a "401 Unauthorized" from the wharf-api then it returns the same response from the provider itself.

I think it would be wise to keep the providers dumb in this way. They do not manage any authentication/authorization at all.

Reply:
As they are separate clients that connect to the untrusted zone which is the wharf-web-client then as far as my understanding goes they would have to individually establish trust with the authorization server.

Alternativly if we start and finish the project to move the provider clients (wharf-provider-github, wharf-provider-azuredevops, wharf-provider-gitlab) into the wharf-api such that they are not exposed microservices then the model could be simplified. 

1) Either they could be microservices that are unauthenticated but ARE restricted to ONLY the trusted network. In this case they could be kept outside the authentication system and just remain unprotected.

2) They could be turned into interfaces which get implemented by the wharf-api endpoint. This way they wouldnt be networked and therefore authentication would not be applicable.

Reply:
While I do like the 2nd alternative more, Go does not have good plugin support. So option 1 is the best for allowing users to add more provider plugins without recompiling the main API. Letting them live as microservices also makes it easier to host them in k8s and docker-compose as then it will be as simple as adding yet another pod/container to the mix and then you're done (in theory).

I'll start working on an RFC for the reworked providers integration.